File: frontend/src/lib/balances.js
/**
 * SPEC §6: balances and the settlement view.
 *
 * Transfers are a computed view. Nothing is stored, nothing can be marked
 * paid, and the app has no concept of an outstanding balance after people
 * start actually sending money.
 */
import { sumCents } from './money.js'

/**
 * Net balance per participant.
 *
 * SPEC §6: (sum of amount_cents on expenses they paid) − (sum of their split
 * rows). Positive means the trip owes them; negative means they owe the trip.
 *
 * @param {Array<{id: string}>} participants in canonical order
 * @param {Array<{paidBy: string, amountCents: number, splits: Array<{participantId: string, amountCents: number}>}>} expenses
 * @returns {Array<{participantId: string, paidCents: number, owedCents: number, netCents: number}>}
 */
export function computeBalances(participants, expenses) {
  const paid = new Map(participants.map((p) => [p.id, 0]))
  const owed = new Map(participants.map((p) => [p.id, 0]))

  for (const expense of expenses) {
    if (paid.has(expense.paidBy)) {
      paid.set(expense.paidBy, paid.get(expense.paidBy) + expense.amountCents)
    }
    for (const split of expense.splits) {
      if (owed.has(split.participantId)) {
        owed.set(split.participantId, owed.get(split.participantId) + split.amountCents)
      }
    }
  }

  return participants.map((p) => ({
    participantId: p.id,
    paidCents: paid.get(p.id),
    owedCents: owed.get(p.id),
    netCents: paid.get(p.id) - owed.get(p.id),
  }))
}

/**
 * Minimised transfer list.
 *
 * SPEC §6: computed greedily — repeatedly match the largest creditor against
 * the largest debtor. Ties fall back to canonical participant order so the
 * same trip always renders the same list.
 *
 * @param {Array<{participantId: string, netCents: number}>} balances in
 *   canonical participant order
 * @returns {Array<{fromParticipantId: string, toParticipantId: string, amountCents: number}>}
 */
export function computeTransfers(balances) {
  const creditors = balances
    .map((b, index) => ({ id: b.participantId, amount: b.netCents, index }))
    .filter((b) => b.amount > 0)
    .sort((a, b) => b.amount - a.amount || a.index - b.index)

  const debtors = balances
    .map((b, index) => ({ id: b.participantId, amount: -b.netCents, index }))
    .filter((b) => b.amount > 0)
    .sort((a, b) => b.amount - a.amount || a.index - b.index)

  const transfers = []
  let c = 0
  let d = 0

  while (c < creditors.length && d < debtors.length) {
    const creditor = creditors[c]
    const debtor = debtors[d]
    const amountCents = Math.min(creditor.amount, debtor.amount)

    if (amountCents > 0) {
      transfers.push({
        fromParticipantId: debtor.id,
        toParticipantId: creditor.id,
        amountCents,
      })
    }

    creditor.amount -= amountCents
    debtor.amount -= amountCents
    if (creditor.amount === 0) c += 1
    if (debtor.amount === 0) d += 1
  }

  return transfers
}

/**
 * SPEC §6: the trip total, alongside balances.
 * @param {Array<{amountCents: number}>} expenses
 * @returns {number}
 */
export function computeTripTotal(expenses) {
  return sumCents(expenses.map((e) => e.amountCents))
}

File: backend/quits/balances.py
"""SPEC §6: balances and the settlement view.

Transfers are a computed view. Nothing is stored, nothing can be marked paid,
and the app has no concept of an outstanding balance after people start
actually sending money.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Balance:
    participant_id: str
    paid_cents: int
    owed_cents: int

    @property
    def net_cents(self) -> int:
        return self.paid_cents - self.owed_cents


@dataclass(frozen=True)
class Transfer:
    from_participant_id: str
    to_participant_id: str
    amount_cents: int


def compute_balances(
    participant_ids: list[str],
    expenses: list[tuple[str, int, list[tuple[str, int]]]],
) -> list[Balance]:
    """Net balance per participant, in the order the participants are given.

    SPEC §6: (sum of amount_cents on expenses they paid) − (sum of their split
    rows). ``expenses`` is a list of ``(paid_by, amount_cents, splits)``.
    """
    paid = dict.fromkeys(participant_ids, 0)
    owed = dict.fromkeys(participant_ids, 0)

    for paid_by, amount_cents, splits in expenses:
        if paid_by in paid:
            paid[paid_by] += amount_cents
        for participant_id, split_amount in splits:
            if participant_id in owed:
                owed[participant_id] += split_amount

    return [Balance(pid, paid[pid], owed[pid]) for pid in participant_ids]


def compute_transfers(balances: list[Balance]) -> list[Transfer]:
    """The minimised transfer list.

    SPEC §6: computed greedily — repeatedly match the largest creditor against
    the largest debtor. Ties fall back to canonical participant order, so the
    same trip always renders the same list.
    """
    creditors = sorted(
        ((b.net_cents, i, b.participant_id) for i, b in enumerate(balances) if b.net_cents > 0),
        key=lambda row: (-row[0], row[1]),
    )
    debtors = sorted(
        ((-b.net_cents, i, b.participant_id) for i, b in enumerate(balances) if b.net_cents < 0),
        key=lambda row: (-row[0], row[1]),
    )

    creditor_amounts = [amount for amount, _, _ in creditors]
    debtor_amounts = [amount for amount, _, _ in debtors]

    transfers: list[Transfer] = []
    c = d = 0
    while c < len(creditors) and d < len(debtors):
        amount = min(creditor_amounts[c], debtor_amounts[d])
        if amount > 0:
            transfers.append(Transfer(debtors[d][2], creditors[c][2], amount))
        creditor_amounts[c] -= amount
        debtor_amounts[d] -= amount
        if creditor_amounts[c] == 0:
            c += 1
        if debtor_amounts[d] == 0:
            d += 1

    return transfers
